### PR TITLE
Don't explode when the client resets a stream while it uploads data.

### PR DESCRIPTION
**Pull request by Lukasa**
See https://twistedmatrix.com/trac/ticket/8682

### CODE REVIEW

File: twisted/web/test/test_http2.py [hunk 1]
```diff
@@ -2079,6 +2079,64 @@ def validate(*args):
...
+        )
+        del frames[-1]
+        frames.append(
+            frameFactory.buildRstStreamFrame(
+                streamID=1, errorCode=h2.errors.INTERNAL_ERROR
+            )
+        )
+
+        requestBytes = frameFactory.clientConnectionPreface()
+        requestBytes += b''.join(f.serialize() for f in frames)
```
Comment (adiroiban) [inline twisted/web/test/test_http2.py:2107]: minor comment. I have seen these 2 lines repeated multiple times. maybe they can be refactored as

```
requestBytes = frameFactory.serializedClientFrames(frames)
```

File: twisted/web/test/test_http2.py [hunk 2]
```diff
@@ -2079,6 +2079,64 @@ def validate(*args):
...
+        frames = framesFromBytes(transport.value())
+
+        # Check that the only WINDOW_UPDATE frames came for the connection.
+        windowUpdateFrames = (
+            f for f in frames
+            if isinstance(f, hyperframe.frame.WindowUpdateFrame)
+        )
+        self.assertTrue(
+            all(frame.stream_id == 0 for frame in windowUpdateFrames)
+        )
```
Comment (adiroiban) [inline twisted/web/test/test_http2.py]: minor comment

Are we receiving a random number of WINDOW_UPDATE frames, and we receive them in a random order?

Can we use something like this?

```
windowUpdateFramesID = (f.stream_id for f in frames if isinstance(f, hyperframe.frame.WindowUpdateFrame))
self.assertEqual([0], windowUpdateFramesID)
```

assertTrue will give as an `false is not true` message, but assertEqual should give a clue if we are not received the expected number or the expected stream id

File: twisted/web/_http2.py [hunk 3]
```diff
@@ -686,7 +686,18 @@ def openStreamWindow(self, streamID, increment):
         @type increment: L{int}
         """
         # TODO: Consider whether we want some kind of consolidating logic here.
```
Comment (adiroiban) [inline twisted/web/_http2.py:688]: I think that is better to have a ticket for this todo and convert it into a FIXME

In the ticket we can add more information, notes and discuss this issue

File: twisted/web/_http2.py [hunk 4]
```diff
@@ -686,7 +686,18 @@ def openStreamWindow(self, streamID, increment):
...
-        self.conn.increment_flow_control_window(increment, stream_id=streamID)
+        try:
+            self.conn.increment_flow_control_window(
+                increment, stream_id=streamID
+            )
+        except h2.exceptions.StreamClosedError:
+            # The stream got reset and we haven't worked it out yet. The stream
+            # window doesn't matter now. We still want to increment the
+            # connection window though.
```
Comment (adiroiban) [inline twisted/web/_http2.py]: `We still want to increment the connection window though.`

dumb question ahead. why ? :)

we want or we should ?

is this behaviour discussed in the h2 rfc?

Comment (Lukasa) [inline twisted/web/_http2.py]: I'll expand this comment.

**Pull request merged**